cpp/include/raft/util/itertools.hpp

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <tuple>
+#include <vector>
+
+/**
+ * Helpers inspired by the Python itertools library
+ *
+ */
+
+namespace raft::util::itertools {
+
+template <class S, typename... Args, size_t... Is>
+inline std::vector<S> product(std::index_sequence<Is...> index, const std::vector<Args>&... vecs)
```

Comment:
Just a side question- is there a reason for explicitly inlining templated functions like this?

Reply:
The point of inlining here was not to avoid linking errors, since it's a templated function as you said. My thinking was just that typically it's only going to be called once for a given set of template arguments (because the use case is mostly defining inputs/outputs for tests/benchmarks).

But since there is little benefit in inlining it anyway, I don't mind removing that keyword.

Reply:
In general, we have been trying to remove the use of inline across the codebase unless absolutely necessary. IIRC, the use of incline (whether host or device) is just a suggestion and doesn't mean the compiler will always honor it, right? But I think if we don't need it then it could help in our quest to decrease compile times.

Reply:
I think if it's something that's not intended to be invoked directly by the user then we should probably at least make an attempt to separate the private bits. If the in-line is absolutely needed then it's fine to keep it

Reply:
On second thought the anonymous namespace wouldn't be appropriate since it's equivalent to making the function static, so I've separated it in raft fashion in a corresponding `detail` namespace, and kept it inline. I've also used the occasion to add documentation to the public API.